subsys/shell/shell_telnet.c

```diff
@@ -0,0 +1,493 @@
+		break;
+	default:
+		cmd->op = NVT_CMD_WONT;
+		break;
+	}
+
+	telnet_command_send_reply((u8_t *)cmd,
+				  sizeof(struct telnet_simple_command));
+}
+
+static void telnet_reply_command(struct telnet_simple_command *cmd)
+{
+	if (!cmd->iac) {
+		return;
+	}
+
+	switch (cmd->op) {
+	case NVT_CMD_AO:
+		/* OK, no output then */
+		sh_telnet->output_lock = true;
+		sh_telnet->line_out.len = 0;
+		k_timer_stop(&sh_telnet->send_timer);
+		break;
+	case NVT_CMD_AYT:
+		telnet_reply_ay_command();
+		break;
+	case NVT_CMD_DO:
+		telnet_reply_do_command(cmd);
+		break;
+	default:
+		LOG_DBG("Operation %u not handled", cmd->op);
+		break;
+	}
+}
+
+#endif /* CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND */
+
+static int telnet_send(void)
+{
+	int err;
+
+	if (sh_telnet->line_out.len == 0) {
+		return 0;
+	}
+
+	err = net_context_send(sh_telnet->client_ctx, sh_telnet->line_out.buf,
+			       sh_telnet->line_out.len, telnet_sent_cb,
+			       K_FOREVER, NULL);
+	if (err < 0) {
+		NET_ERR("Failed to send %d, shutting down", err);
+		telnet_end_client_connection();
+		return err;
+	}
+
+	/* We reinitialize the line buffer */
+	sh_telnet->line_out.len = 0;
+
+	return 0;
+}
+
+static void telnet_send_prematurely(struct k_timer *timer)
+{
+	(void)telnet_send();
+}
+
+static inline bool telnet_handle_command(struct net_pkt *pkt)
+{
+	/* Commands are two or three bytes. */
+	NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(cmd_access, u16_t);
+	struct telnet_simple_command *cmd;
+
+	cmd = (struct telnet_simple_command *)net_pkt_get_data(pkt,
+							       &cmd_access);
+	if (!cmd || cmd->iac != NVT_CMD_IAC) {
+		return false;
+	}
+
+#ifdef CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND
+	LOG_DBG("Got a command %u/%u/%u", cmd->iac, cmd->op, cmd->opt);
+	telnet_reply_command(cmd);
+#endif  /* CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND */
+
+	return true;
+}
+
+static void telnet_recv(struct net_context *client,
+			struct net_pkt *pkt,
+			union net_ip_header *ip_hdr,
+			union net_proto_header *proto_hdr,
+			int status,
+			void *user_data)
+{
+	size_t len;
+
+	if (!pkt || status) {
+		telnet_end_client_connection();
+
+		LOG_DBG("Telnet client dropped (AF_INET%s) status %d",
+			net_context_get_family(client) == AF_INET ?
+			"" : "6", status);
+		return;
+	}
+
+	len = net_pkt_remaining_data(pkt);
+	if (len < TELNET_MIN_MSG) {
+		LOG_DBG("Packet smaller than minimum length");
+		goto unref;
+	}
+
+	if (telnet_handle_command(pkt)) {
+		LOG_DBG("Handled command");
+		goto unref;
+	}
+
+	/* Fifo add */
+	k_fifo_put(&sh_telnet->rx_fifo, pkt);
+
+	sh_telnet->shell_handler(SHELL_TRANSPORT_EVT_RX_RDY,
+				 sh_telnet->shell_context);
+
+	return;
+
+unref:
+	net_pkt_unref(pkt);
+}
+
+static void telnet_accept(struct net_context *client,
+			  struct sockaddr *addr,
+			  socklen_t addrlen,
+			  int error,
+			  void *user_data)
+{
+	if (error) {
+		LOG_ERR("Error %d", error);
+		goto error;
+	}
+
+	if (sh_telnet->client_ctx) {
+		LOG_WRN("A telnet client is already in.");
```

Comment:
Warning sounds too much here, perhaps info is enough as we are just telling user that someone is already connected which does not sound a warning type of thing.

Reply:
> Warning sounds too much here

Well, depends. It may be a security feature too, if a (single) legitimate user tries to log in, and finds that someone else is logged in too.